src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs

```diff
@@ -449,7 +449,8 @@ public static IEnumerable<string> GetAllProjectFileNamesWithXBuild(string soluti
             string solutionFile,
             string msbuildPath)
         {
-            if (RuntimeEnvironmentHelper.IsMono && msbuildPath.Contains("xbuild"))
+            if (RuntimeEnvironmentHelper.IsMono &&
```

Comment:
I don't know if this is the right fix here. This is too late in the process. 

The path should have been discovered correctly in GetMsBuildToolset. 

So if there's a bug, it should be fixed somewhere along that codepath.

Reply:
@nkolev92 I think you didn't understand the fix; `msbuildPath` contains only a path to a folder, while `GetMsbuild(msbuildPath)` contains a path to the binary. That's why the first Contains was not matching: it was /usr/bin, while GetMsbuild(msbuildPath) returns the full path to xbuild.exe, which can then contain "xbuild" string.

Otherwise, I'm not sure what you're requesting.

Reply:
I'll take another look a bit later.

Reply:
So this would work as expected, but the code flow is really weird.
I wouldn't have expected GetMSBuild(string msbuildDirectory) to call GetMSbuild() which looks at the env vars.
In this case that's duplicate work I think. 

I think if that line is removed from that method the tests will still pass.

Reply:
> but the code flow is really weird.

I agree, and I think the culprit is storing paths as strings (a pattern called "primitive obsession"). The consequence of this is that the devs that come to work on this codebase later have a hard time knowing if the path corresponds to a directory or to a file. The long-term fix for this, as I already stated in the PR description, is, IMO, to change all this string-based path variables/fields with DirectoryInfo/FileInfo vars/fields.

However, I was aiming at a quick fix here for now, and once this is accepted, work on the next PR to make the code a bit more decent.

> I think if that line is removed from that method the tests will still pass.

What method? I doubt there are tests for xbuild, are they?

Reply:
Having a bit of a look myself, I see what @knocte's comment in the PR about directories is about. `msbuildPath` always is the path to the directory, not the executable, so I don't see how `msbuildPath.Contains("xbuild")` could ever be true. Without some kind of fix, the call to `GetAllProjectFileNamesWithXBuild` is a dead code path.